tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,644 @@
+    /// the contract that they are the sole owner of the file descriptor they
+    /// are wrapping. Usage of this function could accidentally allow violating
+    /// this contract which can cause memory unsafety in code that relies on it
+    /// being true.
+    ///
+    /// # Errors
+    ///
+    /// This errors if called outside of a [Tokio Runtime] which doesn't have
+    /// [I/O enabled].
+    ///
+    /// [Tokio Runtime]: crate::runtime::Runtime
+    /// [I/O enabled]: crate::runtime::RuntimeBuilder::enable_io
+    unsafe fn try_from_raw_handle(handle: RawHandle) -> io::Result<Self> {
+        let named_pipe = mio::windows::NamedPipe::from_raw_handle(handle);
+
+        Ok(NamedPipe {
+            io: PollEvented::new(named_pipe)?,
+        })
+    }
+
+    /// Enables a named pipe server process to wait for a client process to
+    /// connect to an instance of a named pipe. A client process connects by
+    /// creating a named pipe with the same name.
+    ///
+    /// This corresponds to the [`ConnectNamedPipe`] system call.
+    ///
+    /// [`ConnectNamedPipe`]: https://docs.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-connectnamedpipe
+    ///
+    /// ```no_run
+    /// use tokio::net::windows::NamedPipeBuilder;
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let builder = NamedPipeBuilder::new("\\\\.\\pipe\\mynamedpipe");
+    /// let pipe = builder.create()?;
+    ///
+    /// // Wait for a client to connect.
+    /// pipe.connect().await?;
+    ///
+    /// // Use the connected client...
+    /// # Ok(()) }
+    /// ```
+    pub async fn connect(&self) -> io::Result<()> {
+        loop {
+            match self.io.connect() {
+                Ok(()) => break,
+                Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
+                    self.io.registration().readiness(Interest::WRITABLE).await?;
+                }
+                Err(e) => return Err(e),
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Disconnects the server end of a named pipe instance from a client
+    /// process.
+    ///
+    /// ```no_run
+    /// use tokio::net::windows::NamedPipeBuilder;
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let builder = NamedPipeBuilder::new("\\\\.\\pipe\\mynamedpipe");
+    /// let pipe = builder.create()?;
+    ///
+    /// // Wait for a client to connect.
+    /// pipe.connect().await?;
+    ///
+    /// // Use the pipe...
+    ///
+    /// // Forcibly disconnect the client (optional).
+    /// pipe.disconnect()?;
+    /// # Ok(()) }
+    /// ```
+    pub fn disconnect(&self) -> io::Result<()> {
+        self.io.disconnect()
+    }
+}
+
+impl AsyncRead for NamedPipe {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        unsafe { self.io.poll_read(cx, buf) }
+    }
+}
```

Comment:
I'm guessing this will fail if used on a non-connected server socket.

Reply:
Not if the client is created after the server - then it's just supposed to be connected. All though, there is one weird exception I linked to in another comment that I don't understand which involves custom access:
https://github.com/tokio-rs/tokio/pull/3760/files#diff-697d0148d22589ebb6a5805054e91924230460d8038f36e74d388db00e2a35f0R78